Firestore/core/src/firebase/firestore/local/leveldb_opener.cc

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/local/leveldb_opener.h"
+
+#include <string>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/core/database_info.h"
+#include "Firestore/core/src/firebase/firestore/local/leveldb_persistence.h"
+#include "Firestore/core/src/firebase/firestore/local/local_serializer.h"
+#include "Firestore/core/src/firebase/firestore/remote/serializer.h"
+#include "Firestore/core/src/firebase/firestore/util/filesystem.h"
+#include "Firestore/core/src/firebase/firestore/util/filesystem_detail.h"
+#include "Firestore/core/src/firebase/firestore/util/log.h"
+#include "Firestore/core/src/firebase/firestore/util/path.h"
+#include "Firestore/core/src/firebase/firestore/util/statusor.h"
+#include "Firestore/core/src/firebase/firestore/util/string_format.h"
+#include "absl/strings/match.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+namespace {
+
+using core::DatabaseInfo;
+using remote::Serializer;
+using util::Path;
+using util::RecursivelyCreateDir;
+using util::Status;
+using util::StatusOr;
+using util::StringFormat;
+
+constexpr const char* kReservedPathComponent = "firestore";
+
+Status FromCause(const std::string& message,
+                 const firebase::firestore::util::Status& cause) {
+  if (cause.ok()) return cause;
+
+  return Status(cause.code(), message).CausedBy(cause);
+}
+
+}  // namespace
+
+LevelDbOpener::LevelDbOpener(
+    firebase::firestore::core::DatabaseInfo database_info)
+    : database_info_(std::move(database_info)) {
+}
+
+Path LevelDbOpener::AppDataDir() {
```

Comment:
I feel like it's confusing to have util::AppDataDir() and LevelDbOpener::AppDataDir()...  Perhaps LevelDbOpener::FirestoreAppDataDir() ?